src/plugins/discourse/models.js

```diff
@@ -0,0 +1,43 @@
+// @flow
+
+/**
+ * Represents a normalized Discourse ServerUrl.
+ *
+ * - Protocols http / https are allowed.
+ * - Hostnames are allowed.
+ * - Ports are allowed.
+ * - Other components are not allowed.
+ * - Explicit default ports are removed.
+ * - Trailing slashes are removed.
+ * - URL is lowercased.
+ */
+export opaque type ServerUrl: string = string;
+
+/**
+ * Does a strict validation of serverUrl inputs and returns an opaque type
+ * indicating it's normalized to match our ServerUrl assumptions.
+ */
+export function parseServerUrl(serverUrl: string): ServerUrl {
+  try {
+    const url = new global.URL(serverUrl);
+
+    if (!new RegExp(/^https?/i).test(url.protocol)) {
+      throw "URL should have a http/https protocol";
+    }
+
+    const rebuiltUrl = new global.URL(
+      `${url.protocol}//${url.hostname}:${url.port}`
```

Comment:
FWIW, this matches any protocol _starting_ with `http[s]`:

```javascript
> new RegExp(/^https?/i).test(new global.URL("httpscary://wat").protocol)
true
```

Suggest either `/^https?:$/i` or just `["http:", "https:"].includes`;
the protocol always gets downcased by the URL constructor, anyway, as
your tests correctly show.